### PR TITLE
nonlinear solver

### PR DESCRIPTION
**Pull request by tjhei**
Add a section for NonlinearSolverSelector.

I still need to shorten/remove the description of the nonlinear solver interfaces in the other sections.

### CODE REVIEW

File: 9.5/paper.tex [hunk 1]
```diff
@@ -490,6 +491,41 @@ \subsubsection{Interfaces to Trilinos' Belos and NOX packages}\label{sec:trilino
   elaborate? At least in the KINSOL interfaces, you only update the
   preconditioner whenever setup\_jacobian() is called.}
 
+\subsubsection{Uniform interface for nonlinear solvers}\label{sec:nonlinear}
+
+With the current release, \dealii{} now supports solvers for nonlinear problems
+provided by several different external packages:
+1. KINSOL (part of SUNDIALS), 2. SNES (part of PETSc, see Section~\ref{sec:petsc}), 3. NOX (part of Trilinos, see Section~\ref{sec:trilinos}).
```
Comment (bangerth) [inline 9.5/paper.tex:498]:
```suggestion
(i) KINSOL (part of SUNDIALS); (ii) SNES (part of PETSc, see Section~\ref{sec:petsc}); and (iii) NOX (part of Trilinos, see Section~\ref{sec:trilinos}).
```

File: 9.5/paper.tex [hunk 2]
```diff
@@ -490,6 +491,41 @@ \subsubsection{Interfaces to Trilinos' Belos and NOX packages}\label{sec:trilino
...
+The wrappers are provided by the classes \texttt{PETScWrappers::NonlinearSolver<VectorType>}, \texttt{SUNDIALS::KINSOL<VectorType>}, and \texttt{TrilinosWrappers::NOXSolver<VectorType>}, respectively.
+
+All classes have a very similar interface, except that they vary in the kind of
```
Comment (bangerth) [inline 9.5/paper.tex:501]:
```suggestion
All three of these classes have a very similar interface, except that they vary in the kind of
```

File: 9.5/paper.tex [hunk 3]
```diff
@@ -490,6 +491,41 @@ \subsubsection{Interfaces to Trilinos' Belos and NOX packages}\label{sec:trilino
...
+                                {/*...*/};
+    nonlinear_solver.setup_jacobian = [&](const LA::MPI::Vector &current_u)
+                                      {/*...*/};
+    nonlinear_solver.solve_with_jacobian = [&](const LA::MPI::Vector &rhs,
+                                               LA::MPI::Vector &      dst,
+                                               const double tolerance)
+                                           {/*...*/};
+
+    nonlinear_solver.solve(current_solution);
+\end{c++}
```
Comment (bangerth) [inline 9.5/paper.tex:528]: Will this actually ever use PETSc? It's not a PETSc vector type, and I thought that SNES requires a PETSc vector type.

Comment (tjhei) [reply 9.5/paper.tex:528]: It can use SNES provided the vector type is a PETSc Vector, yes.

**Pull request merged**